asteroid/data/sms_wsj_dataset.py

```diff
@@ -0,0 +1,219 @@
+
+
+class SmsWsjDataset(data.Dataset):
+    """ Dataset class for SMS WSJ source separation.
+
+    Args:
+        json_path (str): The path to the sms_wsj json file.
+        target (str): One of ``'source'``, ``'early'`` or ``'image'``.
+
+            * ``'source'`` non reverberant clean targets signals.
+            * ``'early'`` early reverberation target signals.
+            * ``'image'`` reverberant target signals
+        dset (str): train_si284 for train, cv_dev93 for validation and
+                test_eval92 for test
+        sample_rate (int, optional): The sampling rate of the wav files.
+        segment (float, optional): Length of the segments used for training,
+            in seconds. If None, use full utterances (e.g. for test).
+        single_channel (bool): if True all channels are used if false only
+                    a random channel is used during training
+                    and the first channel during test
+        nondefault_nsrc (int, optional): Number of sources in the training
+            targets.
+        normalize_audio (bool): If True then both sources and the mixture are
+            normalized with the standard deviation of the mixture.
+    """
+    dataset_name = 'SMS_WSJ'
+
+    def __init__(self, json_path, target, dset, sample_rate=8000,
+                 single_channel=True, segment=4.0, nondefault_nsrc=None,
+                 normalize_audio=False):
+        try:
+            import sms_wsj
+        except:
+            import warnings
+            warnings.warn(
+                "Some of the functionality relies on the sms_wsj package "
+                "downloadable from https://github.com/fgnt/sms_wsj ."
+                "The user is encouraged to install the package"
+            )
+        super().__init__()
+        if target not in SMS_TARGETS.keys():
+            raise ValueError('Unexpected task {}, expected one of '
+                             '{}'.format(target, SMS_TARGETS.keys()))
+
+        # Task setting
+        self.json_path = json_path
+        self.target = target
+        self.target_dict = SMS_TARGETS[target]
+        self.single_channel = single_channel
+        self.sample_rate = sample_rate
+        self.normalize_audio = normalize_audio
+        self.seg_len = None if segment is None else int(segment * sample_rate)
+        if not nondefault_nsrc:
+            self.n_src = self.target_dict['default_nsrc']
+        else:
+            assert nondefault_nsrc >= self.target_dict['default_nsrc']
+            self.n_src = nondefault_nsrc
+        self.like_test = self.seg_len is None
+        self.dset = dset
+
+        # Load json files
+
+        from lazy_dataset.database import JsonDatabase
+        db = JsonDatabase(json_path)
+        dataset = db.get_dataset(dset)
+        # Filter out short utterances only when segment is specified
+        orig_len = len(dataset)
+        drop_utt, drop_len = 0, 0
+        if not self.like_test:
```

Comment:
I have also duplicated this code in my WIP Voice Bank + DEMAND impl. So now there are at least 3 duplicates of the code. We should probably merge dataset logic (maybe even preprocessing logic) somehow.

Reply:
If you want to give it a try feel free. At first I didn't want to write abstract classes because there is no _right_ way to do a Dataset class, but if we use it in three places, we could abstract it. 
For the preprocessing, what do you mean exactly?

Reply:
The scripts that create json files. There's some shared code there.